cabal-install/Distribution/Client/Install.hs

```diff
@@ -1447,28 +1456,33 @@ installUnpackedPackage verbosity installLock numJobs
           let testsResult | testsEnabled = TestsOk
                           | otherwise = TestsNotTried
 
-        -- Install phase
-          onFailure InstallFailed $ criticalSection installLock $ do
-            -- Actual installation
-            withWin32SelfUpgrade verbosity uid configFlags
-                                 cinfo platform pkg $ do
-              setup Cabal.copyCommand copyFlags mLogPath
-
-            -- Capture installed package configuration file, so that
-            -- it can be incorporated into the final InstallPlan
-            ipkgs <- genPkgConfs mLogPath
-            let ipkgs' = case ipkgs of
-                            [ipkg] -> [ipkg { Installed.installedUnitId = uid }]
-                            _ -> ipkgs
-            let packageDBs = interpretPackageDbFlags
-                                (fromFlag (configUserInstall configFlags))
-                                (configPackageDBs configFlags)
-            forM_ ipkgs' $ \ipkg' ->
-                registerPackage verbosity comp progdb
-                                packageDBs ipkg'
-                                defaultRegisterOptions
-
-            return (Right (BuildResult docsResult testsResult (find ((==uid).installedUnitId) ipkgs')))
+        -- Benchmarks phase
+          onFailure BenchFailed $ do
+            when (benchmarksEnabled && PackageDescription.hasBenchmarks pkg) $
+                setup Cabal.benchmarkCommand benchmarkFlags' mLogPath
+
+          -- Install phase
+            onFailure InstallFailed $ criticalSection installLock $ do
+              -- Actual installation
+              withWin32SelfUpgrade verbosity uid configFlags
+                                   cinfo platform pkg $ do
+                setup Cabal.copyCommand copyFlags mLogPath
+
```

Comment:
> I can foresee two paths forward here:
> 
> 1. I go back and revert all changes that only involve to `v1-build`, not `v2-build`. I would likely need some guidance for this, since it's not entirely clear to me which code paths are `v1-build`–only. (I didn't realize _this_ code path fell into that category until you told me!)
> 2. I keep this patch as-is. It would make it larger than it strictly has to be, but it would make `--benchmark-options` treated more uniformly across `v1-build` and `v2-build`.

I am not against changing v1-build as such, it should just be a concious desicion. We should get some outside opinions on that since I'm not sure what the plans concerning eventually removing v1-build are. @hvr @phadej @23Skidoo opionions? 

> I did this intentionally, I think. That is, in the original code the whole Install phase block was nested below `onFailure TestsFailed $ do`, so I emulated that design for consistency.

Ah, I didn't realise the indenting here is staircase shaped, it's quite hard to see in the collapsed diff, now that makes sense.

Reply:
Actually I just remembered what I really meant about the v1 thing here. It's totally fine that you're thouching the v1 code since we already had benchmark-options before, so naturally you'd have to adjust things. The thing that rubbed me the wrong way here is that `v1-install` will now suddenly _run_ benchmarks when they are enable when it didn't use to!

So that's a change in behaviour and definetly needs some seperate discussion.

Reply:
> `v1-install` will now suddenly run benchmarks if they are enabled when it didn't use to!

It will? That's surprising to me. I was aware that you could run `cabal v1-install --enable-tests` and it would build the library's test suites alongside the library itself, but I didn't think that it actually ran the tests themselves. (Enabling `--test-show-details=direct` does not reveal any test output, either.) Or is there another way to run test suites with `v1-install`?

I guess what I'm saying is that if the treatment of benchmarks in `v1-install` is strange here, then `v1-install`'s treatment of tests is equally strange. I'm fine with going in either direction, but I'm not even sure how to invoke the strange behavior yet.

Reply:
`--run-tests` is the magic flag you need, so test running is gated against with an additional flag but with your patch all enabled benchmarks will always run.

So you either have to remove the benchmark running stuff from v1-install or add a `--run-benchmarks` flag too.

Reply:
IMO if you want to enable running benchmarks in v1-install that really needs a seperate PR (or at least commit) as it stands now this is pretty sneaky ;)

Reply:
Ah, I was not aware of `--run-tests`! Yes, I certainly agree that this change doesn't make sense unless `--run-benchmarks` exists, and I certainly don't intend to introduce even more flags in this PR. I've gone ahead and reverted this particular change.